graphs/adapter-jgrapht/src/main/kotlin/com/anaplan/engineering/azuki/graphs/adapter/jgrapht/action/AddEdgeToUndirectedGraphAction.kt

```diff
@@ -0,0 +1,23 @@
+package com.anaplan.engineering.azuki.graphs.adapter.jgrapht.action
+
+import com.anaplan.engineering.azuki.graphs.adapter.declaration.action.AddEdgeToUndirectedGraphDeclarableAction
+import com.anaplan.engineering.azuki.graphs.adapter.jgrapht.execution.ExecutionEnvironment
+import org.jgrapht.graph.builder.GraphTypeBuilder
+
+class AddEdgeToUndirectedGraphAction<V>(graphName: String, source: V, target: V) :
+    AddEdgeToUndirectedGraphDeclarableAction<V>(graphName, source, target), JGraphTAction {
+    override fun act(env: ExecutionEnvironment) {
+        env.act(graphName) {
+            if (source == target) {
+                GraphTypeBuilder.forGraph(this).allowingSelfLoops(true).buildGraph()
+            }
```

Comment:
🚫 I presume this is not doing anything, because `act` simply modifies the existing graph object and instead you want to replace it with the one created by `GraphTypeBuilder`. You can check if this is the case easily enough by writing an adapter test for it.

Since this PR is already large enough, I would suggest backing this out and adding a `TODO` comment. Otherwise you'll need to add to `ExecutionEnvironment` a mechanism for replacing an existing graph -- which doesn't seem to exist at the moment -- and call that before this set of actions.